manabi/lock.py

```diff
@@ -94,3 +117,125 @@ def get_lock_list(self, path, *, include_root, include_children, token_only):
+
+    @staticmethod
+    def decode_lock(lock):
+        # Encoding is not needed since a string is also valid for wsgidav
+        if "owner" in lock:
+            owner = lock["owner"]
+            if isinstance(owner, bytes):
+                lock["owner"] = lock["owner"].decode("UTF-8")
+
+    def cursor(self) -> psycopg2.extensions.cursor:
+        return self._connection.cursor()
+
+    def cleanup(self):
+        cursor = self.cursor()
+        cursor.execute("DELETE FROM manabi_lock;")
+
+    def __len__(self):
+        cursor = self.cursor()
+        cursor.execute("SELECT count(*) FROM manabi_lock")
+        return int(cursor.fetchone())
+
+    def __iter__(self):
+        cursor = self.cursor()
+        cursor.execute("SELECT token FROM manabi_lock")
+        for token in cursor.fetchall():
+            yield token[0]
+
+    def __delitem__(self, token):
+        cursor = self.cursor()
+        cursor.execute("DELETE FROM manabi_lock WHERE token = %s", (str(token),))
+
+    def __contains__(self, token):
+        cursor = self.cursor()
+        cursor.execute("SELECT * FROM manabi_lock WHERE token = %s", (str(token),))
+        if cursor.fetchone() is None:
+            return False
+        else:
+            return True
+
+    def __setitem__(self, token, lock):
+        cursor = self.cursor()
+        self.decode_lock(lock)
+        json_lock = json.dumps(lock)
+        cursor.execute(
+            """
+                INSERT INTO manabi_lock(token, data) VALUES (%(token)s, %(data)s)
+                    ON CONFLICT(token) DO
+                    UPDATE SET data = %(data)s WHERE manabi_lock.token = %(token)s
+            """,
+            {
+                "token": token,
+                "data": json_lock,
+            },
+        )
+
+    def __getitem__(self, token):
+        cursor = self.cursor()
+        cursor.execute("SELECT data FROM manabi_lock WHERE token = %s", (str(token),))
+        lock = cursor.fetchone()
+        if lock is None:
+            raise KeyError(f"{token} not found")
+        lock = lock[0]
+        return lock
+
+
+class PostgresReadWriteLock:
+    def __init__(self, connection):
+        connection.commit()
+        connection.autocommit = False
+        connection.set_session(
+            isolation_level=psycopg2.extensions.ISOLATION_LEVEL_SERIALIZABLE
+        )
+        self._connection = connection
+        self._lock = threading.RLock()
+        self.acquire_write = self.acquire_read = self.acquire
+
+    def cursor(self) -> psycopg2.extensions.cursor:
+        return self._connection.cursor()
+
+    def acquire(self):
+        # wsgi-dav accesses the same lock-storage from multiple threads, but postgres only
+        # locks transaction-based and will allow a second lock on the same transaction. So
+        # we have to lock a normal lock and a postgres-lock.
+        self._lock.acquire()
+        cursor = self.cursor()
+        cursor.execute("LOCK TABLE manabi_lock IN ACCESS EXCLUSIVE MODE")
```

Comment:
Still not happy about this one, but I see how it's neccessary to implement the required API.

We *might* get around it by using `SELECT ... FOR UPDATE` on the actual lock entries. Then parallel work in other transactions would still work, as long as it doesn't touch the same records. Not sure if this would be enough for the use case we're facing here though

Reply:
This thing has changed a lot. See the commit  7fd246a8d7c50c75f8c36dbcc145b9128b075664. I may have overdone locking, the locks deeper in the stack in `ManabiPostgresDict` are not strictly needed, but the also don't cost anything, only the semaphore gets incremented/decremented. Let me know what you think. I am ok with remove them.

In theorie if `ManabiPostgresDict` get reused somewhere the locks should stay.